### PR TITLE
DietPi-Software | Beets

### LINKED ISSUE #4434: DietPi-Software | Beets
#### Formal software information
- Software title | 
Beets
- Short description | 
Command-line music organizer
- Official URL (if available) | 
https://beets.io
- GitHub URL (if available) | 
https://github.com/beetbox/beets
- (Official) install documentation | 
https://beets.readthedocs.io/en/stable/guides/main.html

#### Are there similar/alternative software titles available with DietPi-Software?

- Servarr applications

#### What makes your requested software better than the above solutions, if available?

- Doesn't download things, no (built-in) web interface

#### How can DietPi make the installation easier or compatible, than following the install instructions or do APT installation, if available?

- Could make compatible with other music streamers

#### Can you provide the installation steps that you would suggest DietPi-Software to do?

1. Install beets
2. Edit configuration directory
3. Edit config file

#### Are you willing to help maintaining the software installation, e.g. in case of needed setup changes due to updates etc.? This is not needed, but could speed up our decision to implement it, as man power is always a topic :wink:.

- Probably

### PR DESCRIPTION
**Pull request by nonnorm**
**Status**: Ready

**Reference**: https://github.com/MichaIng/DietPi/issues/4434

**Commit list/description**:

- DietPi-Software | Add Beets installation option

### CODE REVIEW

File: dietpi/dietpi-software [hunk 1]
```diff
@@ -652,6 +652,14 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
...
 		# - ARMv6: https://github.com/jellyfin/jellyfin/issues/5011
 		aSOFTWARE_AVAIL_G_HW_ARCH[$software_id,1]=0
+		#------------------
+		software_id=190
+
+		aSOFTWARE_NAME[$software_id]='Beets'
+		aSOFTWARE_DESC[$software_id]='music organizer and manager'
+		aSOFTWARE_CATX[$software_id]=2
+		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/media/#beets'
+		aSOFTWARE_DEPS[$software_id]='130'
```
Comment (MichaIng) [inline dietpi/dietpi-software]: 130 is actually Python pip, while here only the runtime is required. As we install the APT package which has all required Python modules as dependency, I think we can and should remove the dependency here.

It is btw possible to install it via `pip3 install beets`, which brings a newer version and then requires 130. Not sure whether it's worth it, as it needs to be compiled, hence requires the whole build-essentials stack as well.

File: dietpi/dietpi-software [hunk 2]
```diff
@@ -2971,6 +2979,21 @@ _EOF_
 
 		fi
 
+		software_id=190 # Beets
+		if (( ${aSOFTWARE_INSTALL_STATE[$software_id]} == 1 )); then
+
+			Banner_Installing
+
+			G_EXEC mkdir /etc/beets
```
Comment (MichaIng) [inline dietpi/dietpi-software]: As this is not only used for the configuration, but also for the database, what do you think about using `/mnt/dietpi_userdata/beets` to align this with other cases?

FHS would be `/var/lib/beets/library.db`, but IMO it's nice to have everything in userdata, as long as it's not somehow preset by the APT package.

File: dietpi/dietpi-software [hunk 3]
```diff
@@ -2971,6 +2979,21 @@ _EOF_
...
+			G_EXEC chmod 777 /etc/beets
+			
+			G_EXEC eval "echo -e '#!/bin/dash\nexport BEETSDIR=/etc/beets' > /etc/bashrc.d/dietpi-beets.sh"
```
Comment (MichaIng) [inline dietpi/dietpi-software]: No need for the shebang. Using dash is even confusing as `/etc/bashrc.d` of course is only loaded for interactive bash shells. As it is sourced, not executed, no shebang is required. Not sure if we use a shebang in other such cases, but then it should be removed.

File: dietpi/dietpi-software [hunk 4]
```diff
@@ -8215,6 +8238,18 @@ _EOF_
 
 		fi
 
+		software_id=190 # Beets
+		if (( ${aSOFTWARE_INSTALL_STATE[$software_id]} == 1 )); then
+
+			Banner_Configuration
+
+			cat << _EOF_ > /etc/beets/config.yaml
```
Comment (MichaIng) [inline dietpi/dietpi-software]: On reinstall (if the file exists already), we should not overwrite it.

File: dietpi/dietpi-software (hunk 1 above)
Comment (nonnorm) [reply dietpi/dietpi-software]: The Debian package is only 2 minor versions behind, so I think that we'll be fine using that.

File: dietpi/dietpi-software (hunk 2 above)
Comment (nonnorm) [reply dietpi/dietpi-software]: The preset is actually `~/.config/beets`, but the `bashrc.d` script overrides that.

Comment (MichaIng) [reply dietpi/dietpi-software]: Jep, I thought that and makes sense when we expect a single music library on the system 👍.

**Review by MichaIng — DISMISSED** [review]

**Comment by MichaIng** [comment]
A little too geeky for me, to manage a music library from the command line like this, but it works fine.

I changed permissions the way that the library and state files are pre-created with access limited to dietpi user and audio group members. Otherwise, the first user that imports/updates the library will have them blocked with 644 mode for everyone else. Not a big issue, but the way it is now it's a little more defined. The audio files are still created with default mode of course, but no change to do something about that.

**Review by MichaIng — DISMISSED** [review]

**Comment by ravenclaw900** [comment]
Personally I like it because it lets you organize and import music for other apps that don't have an inbuilt music organizer, but will display it using those attributes.

**Comment by MichaIng** [comment]
Jep, I definitely see the use case. For things like this I personally prefer a GUI, but it's great to have a CLI solution in our portfolio 🙂.

**Comment by MichaIng** [comment]
I also just read a bit through the docs and see all the plugins, e.g. to trigger an MPD or Kodi database update automatically, import from IPFS, generate playlists etc etc: https://beets.readthedocs.io/en/stable/plugins/
So it is able to integrate very fine with many other media software, making it quite mighty 👍.

**Review by MichaIng — DISMISSED** [review]

**Comment by MichaIng** [comment]
LGTM. Ready to be merged from your end?

**Comment by ravenclaw900** [comment]
Yep.

**Pull request merged**